### ISSUE TITLE
PipeWriter.GetMemory() InvalidOperationException: Writing is not allowed after writer was completed.

### ISSUE THREAD

**Issue by Kaelum** [dotnet/aspnetcore#26955]
### Describe the bug
Under extreme load, we have noticed that the PipeWriter.GetMemory() method will fail with an InvalidOperationException of "Writing is not allowed after writer was completed."

### To Reproduce
We are processing ~4,000-6,000 requests/second over a TCP connection, for which I have written a custom ConnectionHandler.  The request sizes vary from ~150 bytes to ~100k, but can be replicated with the smaller request size.  For reference, the response is ~3-5 times that of the request.  This issue only started to show up after we were able to run performance tests near the server's maximum capacity.

At first I thought that it was just a closed connection, but these exceptions are occurring while the CancellationToken of ConnectionContext.ConnectionClosed has an IsCancellationRequested property value of false.  It is possible that the connection is closed, but this property is not showing it as such.  Due to the way that the loss of a connection works, it creates a race condition and we have not had any success at reproducing the issue in a debug session.

### Exceptions (if any)
InvalidOperationException: Writing is not allowed after writer was completed.

### Further technical details
- ASP.NET Core version: 3.1.301
- This is running on AWS Linux, latest release, netcoreapp3.1, Self-Contained, linux-x64

**Comment by davidfowl** [comment]
We need to see the code. Can you share a repro with us?

**Comment by Kaelum** [comment]
It will take me a few days to get you an application to replicate this, but I can share the method where it is occuring:

```C#
private async Task SendResponseAsync(
	PipeWriter writer,
	HttpResult httpResult,
	bool isHttpRequest,
	CancellationToken closedToken
)
{
	if (closedToken.IsCancellationRequested)
	{
		return;
	}

	int totalBytes;
	Memory<byte> outputMemory;

	if (isHttpRequest)
	{
		string httpResponseHeader = string.Format(
			HTTP_RESPONSE_FORMAT,
			httpResult.HttpStatusCode.ToString("D"),
			HttpStatusDescription.Get(httpResult.HttpStatusCode),
			httpResult.ContentType,
			httpResult.Response.Length
		);

		byte[] httpResponseHeaderBytes = Encoding.UTF8.GetBytes(httpResponseHeader);
		totalBytes = httpResult.Response.Length + httpResponseHeaderBytes.Length;

		if (closedToken.IsCancellationRequested)
		{
			return;
		}

		outputMemory = writer.GetMemory(totalBytes);
		httpResponseHeaderBytes.CopyTo(outputMemory);
		httpResult.Response.CopyTo(outputMemory.Slice(httpResponseHeaderBytes.Length));
	}
	else
	{
		totalBytes = httpResult.Response.Length;

		if (closedToken.IsCancellationRequested)
		{
			return;
		}

		outputMemory = writer.GetMemory(totalBytes);
		httpResult.Response.CopyTo(outputMemory);
	}

	if (closedToken.IsCancellationRequested)
	{
		return;
	}

	writer.Advance(totalBytes);
	await writer.FlushAsync();

#if DEBUG
	_logger.LogDebug(
		"outputMemory {0} bytes ({1} bytes requested)",
		outputMemory.Span.Length,
		totalBytes
	);
#endif
}
```

I am using the following to filter out the exceptions as defined by this issue.  When using a DEBUG build, it always logs "false":

```C#
catch (InvalidOperationException ex2) when ("Writing is not allowed after writer was completed.".Equals(ex2.Message, StringComparison.InvariantCultureIgnoreCase))
{
#if DEBUG
	_logger.LogWarning("{0} (closed: {1})", ex.Message, closedToken.IsCancellationRequested);
#endif
}
```

The **GetMemory** and **Advance** methods are periodically throwing exceptions that should not be possible.  Here are some new exceptions that we discovered today:

```
System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values. (Parameter 'bytes')
   at System.IO.Pipelines.ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument argument)
   at System.IO.Pipelines.Pipe.Advance(Int32 bytes)
   at Webroot.Bcap.Connection.BcapApiTcpHandler.SendResponseAsync(PipeWriter writer, LogResult logResult, Boolean isHttpRequest, CancellationToken closedToken) in /builds/SD3/BCAP/src/Webroot.Bcap.Core/Connection/BcapApiTcpHandler.cs:line 689
   at Webroot.Bcap.Connection.BcapApiTcpHandler.<>c__DisplayClass32_2.<<OnConnectedAsync>g__sendResponseAsync|3>d.MoveNext() in /builds/SD3/BCAP/src/Webroot.Bcap.Core/Connection/BcapApiTcpHandler.cs:line 422
--- End of stack trace from previous location where exception was thrown ---
   at Webroot.Bcap.Processing.BcapApiProcessor.ProcessRequestAsync[TMethod](String serviceType, IPAddress remoteIpAddress, CreateBcapRequestAsync`1 createBcapRequestAsync, SendResponseAsync sendResponseAsync) in /builds/SD3/BCAP/src/Webroot.Bcap.Core/Processing/BcapApiProcessor.cs:line 382
```

```
System.ArgumentException: Destination is too short. (Parameter 'destination')
   at Webroot.Bcap.Connection.BcapApiTcpHandler.SendResponseAsync(PipeWriter writer, LogResult logResult, Boolean isHttpRequest, CancellationToken closedToken) in /builds/SD3/BCAP/src/Webroot.Bcap.Core/Connection/BcapApiTcpHandler.cs:line 689
   at Webroot.Bcap.Connection.BcapApiTcpHandler.<>c__DisplayClass32_2.<<OnConnectedAsync>g__sendResponseAsync|3>d.MoveNext() in /builds/SD3/BCAP/src/Webroot.Bcap.Core/Connection/BcapApiTcpHandler.cs:line 422
--- End of stack trace from previous location where exception was thrown ---
   at Webroot.Bcap.Processing.BcapApiProcessor.ProcessRequestAsync[TMethod](String serviceType, IPAddress remoteIpAddress, CreateBcapRequestAsync`1 createBcapRequestAsync, SendResponseAsync sendResponseAsync) in /builds/SD3/BCAP/src/Webroot.Bcap.Core/Processing/BcapApiProcessor.cs:line 382	
```

The last one is also demonstrating a StackTrace failure, as the line that is throwing that exception is one or both of the following lines:

- httpResponseHeaderBytes.CopyTo(outputMemory);
- httpResult.Response.CopyTo(outputMemory);

All of these exceptions are random, with these last 2 occurring approximate once every few minutes.  I you still need a project, I can continue to work on one as time permits.

**Comment by BrennanConroy** [comment]
We can't figure out how the exceptions are occurring from this method alone unfortunately. Looking forward to a repro app.

**Comment by Kaelum** [comment]
@BrennanConroy we've learned a lot more since I originally posted this, which has lead us to asking the question in https://github.com/dotnet/core/issues/5559.  At the time that I originally posted this, all testing was only being performed on AWS Linux instances.  We learned that using threads in addition to the thread that the connection opened with, would cause strange timing/race conditions even if we synchronized the threads, which lead to Heartbeat warnings and application crashes.

Last week I ran the same tests, with the exception of the tests for the topic of this issue, for 6 hours on a Windows Server 2016 VM and everything ran perfectly.  AWS Linux is guaranteed to fail within 1 hour, and will start logging errors within 20 minutes, if not sooner.  Once we get an official response to https://github.com/dotnet/core/issues/5559 we'll know how to move forward.  But at this time, it looks like .NET Core does not support AWS Linux.

**Comment by davidfowl** [comment]
I am doubt it has anything to do with amazon Linux. It looks like a hard to track down race condition. For us to help here we'd need something runnable that can reproduce the problem and time to isolated it. Without that it'll be a game of educated guesses back and forth.

**Comment by Kaelum** [comment]
@davidfowl I received a reply on the other thread, so to cover our bases, we tested Ubuntu 20.04 and have the same issue.  However, our current issue is not the issue that this thread was opened under, as we created a work-around for this issue by removing all additional threads and proceeding serially.  So, I'll probably open a new issue for what we are currently observing.  The issues could all be 100% related, but we don't know at this point.  Here is a little background on the current issues that I'm create a new issue for:

- They only occur on Linux, Windows Server 2016 does not exhibit any of the the issues under any load.
- The issues do not occur if we max out the throughput over 100 concurrent connections for a total throughput of 6,000 req/sec,  All connections are reused and only closed at the end of the test.
- The issues do occur if we test with ~1,000+ concurrent connections for any throughput beyond 15% of above, same conditions.
- We are using the StackExchange.Redis library to access a Redis database.  99.999% of the operations are MGET.

I don't know if we're going to be able to replicate the issues w/o Redis, which is why I hadn't been able to create an application for you to test.  I'll try replacing the StackExchange.Redis calls with random delays and see if it still happens, but I don't think that it will.  The StackExchange.Redis library uses Pipelines as well, so the issues probably lie somewhere in them.
